app/lib/utils.ts

```diff
@@ -70,4 +71,73 @@ export class CountingSemaphore {
     const resolve = this.queue.shift();
     resolve?.();
   }
-}
\ No newline at end of file
+}
+
+export function buildSeriesMap(games: GamesTable[]) {
+  const gamesMap = new Map<number, GamesTable>();
+  const prequelsMap = new Map<number, GamesTable[]>(); // game may have multiple sequels
+  const seriesMap = new Map<number, GamesTable[]>(); // series will likely have multiple entries
+  const assignedGames = new Set<number>(); // to keep track of roots that may show up in other series (does not need to be accessed with get, so is a set)
+
+  // construct maps of games and prequels
+  games.forEach((game) => {
+    // construct game map for future lookups
+    gamesMap.set(game.game_id, game);
+    // construct prequel map if the game has a prequel, else it will likely be a root (but these are not the only roots)
+    if (game.prequel_id) {
+      if (!prequelsMap.has(game.prequel_id)) {
+        prequelsMap.set(game.prequel_id, []); // if prequel not in prequelsMap, add a new entry with the prequel as the key
+      }
+      // in either case, add the sequel to the prequelsMap associated to the prequel key
+      prequelsMap.get(game.prequel_id)!.push(game);
+    }
+  });
+
+  // construct array of root games of series
+  const roots = games.filter(
+    (game) =>
+      game.tried === null && // game must be unplayed to be a root, or it won't be shown
+      // game must also either: not have a prequel, or have a prequel that does not exist in the list of games (i.e. was deleted), or have a prequel that was played (to be removed later if in middle of series)
+      (!game.prequel_id ||
+        !gamesMap.has(game.prequel_id) ||
+        gamesMap.get(game.prequel_id)?.tried !== null),
+  );
+
+  // recursive function to build the series chain, called below in production of seriesMap
+  function buildSeriesChain(game: GamesTable, series: GamesTable[]) {
+    // add every game to the assigned games so it cannot show up as a root (in the future, past covered later)
+    assignedGames.add(game.game_id);
+    // add the game to the series chain only if unplayed (not relevant for root, but for the recursive cases)
+    if (game.tried === null) {
+      series.push(game);
+    }
+    // if the game is a prequel and is therefore in prequelsMap:
+    if (prequelsMap.has(game.game_id)) {
+      // go to the sequels and continue building the chain for each of them
+      prequelsMap.get(game.game_id)!.forEach((sequel) => {
+        // if the game is already in seriesMap as a root, then we need to delete that series
+        if (seriesMap.has(game.game_id)) {
+          seriesMap.delete(game.game_id);
+        }
```

Comment:
```suggestion
        // Ensure that we aren't building a duplicate sub range of the same series
        seriesMap.delete(game.game_id);
```

We can skip the `.has` since `.delete` will do that for us internally